primitives/src/contracts_registry.rs

```diff
@@ -69,6 +72,42 @@ pub struct RegistryContract<Hash, AccountId, BalanceOf, BlockNumber> {
     pub meta: ContractMetadata,
 }
 
+pub trait KindValidator {
+    fn can_instantiate(&self) -> bool;
+    fn can_generate_side_effects(&self) -> bool;
+    fn can_remunerate(&self) -> bool;
+    fn has_storage(&self) -> bool;
+}
+
+impl<Hash: Encode, AccountId: Encode, BalanceOf: Encode, BlockNumber: Encode> KindValidator
+    for RegistryContract<Hash, AccountId, BalanceOf, BlockNumber>
+{
+    fn can_instantiate(&self) -> bool {
+        match self.meta.get_contract_type() {
+            ContractType::System => false,
+            _ => true,
+        }
+    }
+
+    fn can_generate_side_effects(&self) -> bool {
+        match self.meta.get_contract_type() {
+            ContractType::VanillaWasm | ContractType::VanillaEvm => false,
+            _ => true,
+        }
+    }
+
+    fn can_remunerate(&self) -> bool {
+        match self.meta.get_contract_type() {
+            ContractType::VanillaWasm | ContractType::VanillaEvm => false,
+            _ => true,
+        }
+    }
+
+    fn has_storage(&self) -> bool {
+        todo!()
```

Comment:
todo here, is this planned for this ticket?

Reply:
I'll push the changes for this one in a minute. But I'll open a separate ticket for the 3vm counterpart